File: packages/electron-compat/src/menu.ts
import { host } from "./host.js";

export type MenuItemRole =
  | "quit" | "close" | "minimize" | "hide" | "hideOthers" | "unhide"
  | "about" | "services" | "cut" | "copy" | "paste" | "selectAll"
  | "undo" | "redo" | "togglefullscreen";

export type MenuItemType = "normal" | "separator" | "submenu" | "checkbox" | "radio";

export interface MenuItemConstructorOptions {
  id?: string;
  label?: string;
  type?: MenuItemType;
  role?: MenuItemRole | string;
  accelerator?: string;
  enabled?: boolean;
  submenu?: MenuItemConstructorOptions[] | Menu;
  click?: (menuItem: MenuItem) => void;
}

const ROLE_ACCELERATORS: Record<string, string> = {
  quit: "CmdOrCtrl+Q",
  close: "CmdOrCtrl+W",
  minimize: "CmdOrCtrl+M",
  hide: "CmdOrCtrl+H",
  hideOthers: "CmdOrCtrl+Alt+H",
  cut: "CmdOrCtrl+X",
  copy: "CmdOrCtrl+C",
  paste: "CmdOrCtrl+V",
  selectAll: "CmdOrCtrl+A",
  undo: "CmdOrCtrl+Z",
  redo: "CmdOrCtrl+Shift+Z",
  togglefullscreen: "Ctrl+Cmd+F",
};

function normaliseAccel(a: string): string {
  return a
    .split("+")
    .map((p) => p.trim())
    .filter(Boolean)
    .map((p) => {
      const l = p.toLowerCase();
      if (l === "command" || l === "cmd" || l === "cmdorctrl" || l === "commandorcontrol") return "cmdorctrl";
      if (l === "control" || l === "ctrl") return "cmdorctrl";
      if (l === "option" || l === "alt") return "alt";
      if (l === "shift") return "shift";
      return l;
    })
    .sort()
    .join("+");
}

let nextId = 0;
const clickHandlers = new Map<string, (mi: MenuItem) => void>();
const idToItem = new Map<string, MenuItem>();
let handlerInstalled = false;

function ensureHandler() {
  if (handlerInstalled) return;
  handlerInstalled = true;
  host().on("menu.click", (msg: { id: string }) => {
    const cb = clickHandlers.get(msg.id);
    const mi = idToItem.get(msg.id);
    if (cb && mi) cb(mi);
  });
}

export class MenuItem {
  readonly id: string;
  label: string;
  type: MenuItemType;
  role?: string;
  accelerator?: string;
  enabled: boolean;
  submenu?: Menu;
  click?: (mi: MenuItem) => void;

  constructor(options: MenuItemConstructorOptions) {
    this.id = options.id ?? `mi_${++nextId}`;
    this.label = options.label ?? "";
    this.type = options.type ?? (options.submenu ? "submenu" : "normal");
    this.role = options.role as string | undefined;
    this.accelerator = options.accelerator;
    this.enabled = options.enabled ?? true;
    this.click = options.click;
    if (options.submenu) {
      this.submenu = options.submenu instanceof Menu
        ? options.submenu
        : Menu.buildFromTemplate(options.submenu);
    }
    if (this.click) {
      clickHandlers.set(this.id, this.click);
      idToItem.set(this.id, this);
    }
  }

  toWire(): unknown {
    return {
      id: this.id,
      label: this.label,
      type: this.type,
      role: this.role,
      accelerator: this.accelerator,
      enabled: this.enabled,
      submenu: this.submenu ? this.submenu.items.map((i) => i.toWire()) : [],
    };
  }
}

function collectAccelerators(items: MenuItem[], out: Map<string, string>) {
  for (const it of items) {
    const claim = it.role ? ROLE_ACCELERATORS[it.role] : it.accelerator;
    if (claim) {
      const key = normaliseAccel(claim);
      const owner = it.role ? `role: '${it.role}'` : `label: '${it.label}'`;
      const prev = out.get(key);
      if (prev && prev !== owner) {
        throw new Error(
          `Menu accelerator collision: '${claim}' is claimed by both ${prev} and ${owner}. ` +
          `On macOS this raises an NSException and crashes the host. Pick a different accelerator.`,
        );
      }
      out.set(key, owner);
    }
    if (it.submenu) collectAccelerators(it.submenu.items, out);
  }
}

export class Menu {
  readonly items: MenuItem[] = [];

  append(item: MenuItem): void {
    this.items.push(item);
  }

  static buildFromTemplate(template: MenuItemConstructorOptions[]): Menu {
    const menu = new Menu();
    for (const opts of template) menu.append(new MenuItem(opts));
    collectAccelerators(menu.items, new Map());
    return menu;
  }

  static async setApplicationMenu(menu: Menu | null): Promise<void> {
    ensureHandler();
    const template = menu ? menu.items.map((i) => i.toWire()) : [];
    await host().request("menu.setApplicationMenu", { template });
  }

  static getApplicationMenu(): Menu | null {
    return null;
  }
}

File: crates/volt-core/src/menu.rs
// SPDX-License-Identifier: Apache-2.0

use anyhow::Result;
use muda::{
    accelerator::Accelerator, AboutMetadata, Menu, MenuItem, PredefinedMenuItem, Submenu,
};
use std::str::FromStr;

use crate::protocol::MenuItemSpec;

pub fn build_menu(template: &[MenuItemSpec]) -> Result<Menu> {
    let menu = Menu::new();
    for spec in template {
        attach(&menu, spec)?;
    }
    Ok(menu)
}

pub fn build_submenu_as_menu(template: &[MenuItemSpec]) -> Result<Menu> {
    build_menu(template)
}

fn attach(parent: &Menu, spec: &MenuItemSpec) -> Result<()> {
    if !spec.submenu.is_empty() {
        let sub = Submenu::new(spec.label.as_deref().unwrap_or(""), spec.enabled.unwrap_or(true));
        for child in &spec.submenu {
            attach_sub(&sub, child)?;
        }
        parent.append(&sub)?;
    } else {
        parent.append(&*item(spec)?)?;
    }
    Ok(())
}

fn attach_sub(parent: &Submenu, spec: &MenuItemSpec) -> Result<()> {
    if !spec.submenu.is_empty() {
        let sub = Submenu::new(spec.label.as_deref().unwrap_or(""), spec.enabled.unwrap_or(true));
        for child in &spec.submenu {
            attach_sub(&sub, child)?;
        }
        parent.append(&sub)?;
    } else {
        parent.append(&*item(spec)?)?;
    }
    Ok(())
}

fn item(spec: &MenuItemSpec) -> Result<Box<dyn muda::IsMenuItem>> {
    if spec.kind.as_deref() == Some("separator") {
        return Ok(Box::new(PredefinedMenuItem::separator()));
    }
    if let Some(role) = spec.role.as_deref() {
        return Ok(predefined(role, spec.label.as_deref()));
    }
    let label = spec.label.as_deref().unwrap_or("");
    let enabled = spec.enabled.unwrap_or(true);
    let accel = spec
        .accelerator
        .as_deref()
        .and_then(|s| Accelerator::from_str(s).ok());
    let id = spec.id.clone().unwrap_or_default();
    Ok(Box::new(MenuItem::with_id(id, label, enabled, accel)))
}

fn predefined(role: &str, label: Option<&str>) -> Box<dyn muda::IsMenuItem> {
    let l = label;
    match role {
        "quit" => Box::new(PredefinedMenuItem::quit(l)),
        "close" | "closeWindow" => Box::new(PredefinedMenuItem::close_window(l)),
        "minimize" => Box::new(PredefinedMenuItem::minimize(l)),
        "hide" => Box::new(PredefinedMenuItem::hide(l)),
        "hideOthers" => Box::new(PredefinedMenuItem::hide_others(l)),
        "unhide" | "showAll" => Box::new(PredefinedMenuItem::show_all(l)),
        "about" => Box::new(PredefinedMenuItem::about(l, Some(AboutMetadata::default()))),
        "services" => Box::new(PredefinedMenuItem::services(l)),
        "separator" => Box::new(PredefinedMenuItem::separator()),
        "cut" => Box::new(PredefinedMenuItem::cut(l)),
        "copy" => Box::new(PredefinedMenuItem::copy(l)),
        "paste" => Box::new(PredefinedMenuItem::paste(l)),
        "selectAll" | "selectall" => Box::new(PredefinedMenuItem::select_all(l)),
        "undo" => Box::new(PredefinedMenuItem::undo(l)),
        "redo" => Box::new(PredefinedMenuItem::redo(l)),
        "togglefullscreen" | "toggleFullscreen" => Box::new(PredefinedMenuItem::fullscreen(l)),
        _ => Box::new(MenuItem::new(l.unwrap_or(role), true, None)),
    }
}
